### PR TITLE
reorganize Flax/JAX stack in 2023a: move `jax` + `Optax` to `gfbf/2023a` toolchain + use standalone `Flax` + `absl-py` as dependencies

### PR DESCRIPTION
**Pull request by lexming**
(created using `eb --new-pr`)

Adding new easyconfig for `Flax`, which deserves its own package; and for `absl-py`, which is used in many places already.

This clarifies the dependency tree of the Flax/JAX stack as:
`absl-py` > `jax` > `Optax` > `Flax`

Changelog:
* move `jax` v0.4.25 from foss/2023a to gfbf/2023a following the changes introduced in #20119
* simplify easyconfig of `jax` v0.4.25:
  * replace component on `absl-py` with regular dependency
  * define `source_urls` individually per each source sownloaded
  * add again the test step in jax
* move `Optax` v0.2.2 from foss/2023a to gfbf/2023a and remove redundant extensions already provided by its dependencies
* add easyconfig for `Flax` v0.8.4
* remove redundant extensions from `scvi-tools` already provided by its dependencies

### CODE REVIEW

**Comment by verdurin** [comment]
Test report by @verdurin
**SUCCESS**
Build succeeded for 35 out of 35 (5 easyconfigs in total)
easybuild-el8.cloud.in.bmrc.ox.ac.uk - Linux Rocky Linux 8.10, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 3.6.8
See https://gist.github.com/verdurin/348867169b5e2ae44f1bafbd338210db for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
**FAILED**
Build succeeded for 5 out of 6 (5 easyconfigs in total)
node603.hydra.os - Linux Rocky Linux 8.10, x86_64, AMD EPYC 9384X 32-Core Processor (x86_64_v4), Python 3.6.8
See https://gist.github.com/lexming/7ceda0ef8964ff32ebd742e77dcca2af for a full test report.

**Comment by yqshao** [comment]
Test report by @yqshao
**FAILED**
Failed during parsing of the easyconfigs, so no ecs were built (1 easyconfigs in total)
alvis1-05 - Linux Rocky Linux 8.9, x86_64, Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz, 1 x NVIDIA Tesla V100-SXM2-32GB, 550.54.14, Python 3.6.8
See https://gist.github.com/yqshao/529a93402978367a38c4d29825fb2163 for a full test report.

**Comment by yqshao** [comment]
Test report by @yqshao
**FAILED**
Failed during parsing of the easyconfigs, so no ecs were built (1 easyconfigs in total)
alvis1-08 - Linux Rocky Linux 8.9, x86_64, Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz, 1 x NVIDIA Tesla V100-SXM2-32GB, 550.54.14, Python 3.6.8
See https://gist.github.com/yqshao/60325e0073671a7b0616d98c4b5775b2 for a full test report.

**Comment by yqshao** [comment]
Test report by @yqshao
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
alvis1-09 - Linux Rocky Linux 8.9, x86_64, Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz, 1 x NVIDIA Tesla V100-SXM2-32GB, 550.54.14, Python 3.6.8
See https://gist.github.com/yqshao/66a542e9454726eab1a97d6ec9916587 for a full test report.

**Comment by boegel** [comment]
@boegelbot: please test @ generoso

**Comment by yqshao** [comment]
Test report by @yqshao
**FAILED**
Build succeeded for 8 out of 15 (5 easyconfigs in total)
vera13-1 - Linux Rocky Linux 8.9, x86_64, Intel(R) Xeon(R) Gold 6130 CPU @ 2.10GHz, Python 3.6.8
See https://gist.github.com/yqshao/5c4787ce824f1ae76618aec2ea300543 for a full test report.

**Comment by lexming** [comment]
@boegel Test in generoso failed due to a dangling lock

> == 2024-07-31 14:51:30,027 build_log.py:171 ERROR EasyBuild crashed with an error (at easybuild/easybuild-framework/easybuild/base/exceptions.py:126 in __init__): Lock /project/boegelbot/Rocky8/haswell/software/.locks/_project_boegelbot_Rocky8_haswell_software_jax_0.4.25-gfbf-2023a.lock already exists, aborting! (at easybuild/easybuild-framework/easybuild/tools/filetools.py:2013 in check_lock)

**Comment by lexming** [comment]
Test report by @lexming
**FAILED**
Build succeeded for 5 out of 6 (5 easyconfigs in total)
node613.hydra.os - Linux Rocky Linux 8.10, x86_64, AMD EPYC 9384X 32-Core Processor (x86_64_v4), Python 3.6.8
See https://gist.github.com/lexming/a836ba15d3115ff5b179c47daa1e83cb for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
**SUCCESS**
Build succeeded for 5 out of 5 (5 easyconfigs in total)
node605.hydra.os - Linux Rocky Linux 8.10, x86_64, AMD EPYC 9384X 32-Core Processor (x86_64_v4), Python 3.6.8
See https://gist.github.com/lexming/c1364a7156454ddc04b7be9b7b334e6a for a full test report.

File: easybuild/easyconfigs/o/Optax/Optax-0.2.2-gfbf-2023a.eb [hunk 1]
```diff
@@ -20,13 +20,6 @@ exts_list = [
     ('toolz', '0.12.1', {
         'checksums': ['ecca342664893f177a13dac0e6b41cbd8ac25a358e5f215316d43e2100224f4d'],
     }),
-    ('flit_core', '3.9.0', {
-        'checksums': ['72ad266176c4a3fcfab5f2930d76896059851240570ce9a98733b658cb786eba'],
-    }),
-    ('absl-py', '2.1.0', {
```
Comment (boegel) [inline easybuild/easyconfigs/o/Optax/Optax-0.2.2-gfbf-2023a.eb:26]: @lexming Why isn't absl-py` required as dependency if this is removed? Was that incorrect to begin with?

**Comment by boegel** [comment]
> @boegel Test in generoso failed due to a dangling lock

I've cleaned up the lock, trying again...

**Comment by boegel** [comment]
@boegelbot please test @ generoso
CORE_CNT=16

**Comment by boegel** [comment]
@boegelbot please test @ jsc-zen3
CORE_CNT=16

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 4 out of 5 (5 easyconfigs in total)
node3105.skitty.os - Linux RHEL 8.8, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/boegel/734a90fdf5b586bb5168fa2c1d36f79b for a full test report.

**Comment by boegel** [comment]
test report failed for me because the existing `jax/0.4.25-foss-2023a` was picked up, I'll remove that and try again...

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 5 out of 6 (5 easyconfigs in total)
node3105.skitty.os - Linux RHEL 8.8, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/boegel/0e9cd00e712862648fb3277e8f7514a4 for a full test report.

**Comment by lexming** [comment]
@boegel you test failed because your _robots_ path has both `jas-0.4.25-gfbf-2023a` from this PR and `jas-0.4.25-foss-2023a` from your local file tree. So `scvi-tools` picks `scvi-tools-1.1.2-foss-2023a.eb` instead of using jax from this PR.

**Comment by boegel** [comment]
> @boegel you test failed because your _robots_ path has both `jax-0.4.25-gfbf-2023a` from this PR and `jax-0.4.25-foss-2023a` from your local file tree. So `scvi-tools-1.1.2-foss-2023a` picks `jax-0.4.25-foss-2023a` instead of using jax from this PR.

Indeed. I'm a bit confused why this is not a problem for the bots though (since they use `develop` branch to test PRs with, which has `jax-0.4.25-foss-2023a.eb` still in there currently.

**Comment by branfosj** [comment]
> > @boegel you test failed because your _robots_ path has both `jax-0.4.25-gfbf-2023a` from this PR and `jax-0.4.25-foss-2023a` from your local file tree. So `scvi-tools-1.1.2-foss-2023a` picks `jax-0.4.25-foss-2023a` instead of using jax from this PR.
> 
> Indeed. I'm a bit confused why this is not a problem for the bots though (since they use `develop` branch to test PRs with, which has `jax-0.4.25-foss-2023a.eb` still in there currently.

Because you have `LMOD_DISABLE_NAME_AUTOSWAP` set? Whereas the bots do not and will just silently do the switch between the possible modules.

**Comment by boegel** [comment]
> > > @boegel you test failed because your _robots_ path has both `jax-0.4.25-gfbf-2023a` from this PR and `jax-0.4.25-foss-2023a` from your local file tree. So `scvi-tools-1.1.2-foss-2023a` picks `jax-0.4.25-foss-2023a` instead of using jax from this PR.
> > 
> > 
> > Indeed. I'm a bit confused why this is not a problem for the bots though (since they use `develop` branch to test PRs with, which has `jax-0.4.25-foss-2023a.eb` still in there currently.
> 
> Because you have `LMOD_DISABLE_NAME_AUTOSWAP` set? The bots do not and will swap between the possible modules.

Exactly, I overlooked that part. You're 100% right.

I'll dance around this using `--robot-paths /tmp/$USER` and get a new test report uploaded.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 5 out of 5 (5 easyconfigs in total)
node3123.skitty.os - Linux RHEL 8.8, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/boegel/f1721d4bc981b2dddd8cf6082f14bcc9 for a full test report.

**Review by boegel — APPROVED** [review]
lgtm

**Comment by boegel** [comment]
Going in, thanks @lexming!

**Pull request merged**